### DISCUSSION TITLE
What do you want from OpenMage to make it competitive for the future?

### DISCUSSION THREAD

**Discussion by Flyingmana** [OpenMage/magento-lts#2873] [General]
The year ends, a new one starts.
We got stability, continuity and compatibility(with modern php/sql) covered.

Now we know the various things our "competitors" are up to.
So looking at them, what are the features which are most interesting for you, or make a relevant difference to change to them because we dont have them(or not good enough)

**Comment by elidrissidev** [comment]
Some features that I'd love to see in OM but also think will bring a lot of value to users:

- Multi-stock functionality. Currently everything is hardcoded around the "Default" stock, but at least the base is there, so I think it's feasible.
- Audit logs. If relying on the Admin panel alone as a way to manage a store, you'll often play the detective role trying to figure out who did what.
- Editing an Order without canceling it and creating a new one.

**Reply by sreichel** [reply]
>  Audit logs. If relying on the Admin panel alone as a way to manage a store, you'll often play the detective role trying to figure out who did what.

Something like https://github.com/firegento/firegento-adminmonitoring?

**Reply by elidrissidev** [reply]
Yup, that's a good starting point.

**Reply by fballiano** [reply]
I didn't know that module but seems very nice, we should add it to https://openmage.readthedocs.io/en/latest/modules/index.html

actually... that readthedocs website should be completely removed, it's not easy to navigate and it's mostly empty, we should just have a "awesome openmage" modules list in a wiki page (here on the repo) or in .MD file here on the main repo.

**Reply by elidrissidev** [reply]
I would prefer having a repository called "awesome-openmage" in which we curate the best modules that are still maintained. Then we can add it to [sindresorhus/awesome](https://github.com/sindresorhus/awesome) for more recognition.

**Reply by fballiano** [reply]
sometime ago I created https://github.com/fballiano/awesome-openmage

**Reply by sreichel** [reply]
:+1: 

Very first ... "awesome openmage" should be linked and [updated](https://github.com/flagbit/Magento-ChangeAttributeSet).

There are many missing feature and still a lot of extenson, so it difficult to tell whats "needed".

From my programmers view there is one PR that i want to have merge sometimes ... #423. Every model (and every block-class - another PR) could have its own event-prefix, to be accessed by observers easily ...

**Reply by fballiano** [reply]
> Very first ... "awesome openmage" should be linked and [updated](https://github.com/flagbit/Magento-ChangeAttributeSet).

I started as a test because I couldn't merge things into our "readthedoc", but I didn't mean it to be already "complete" :-) But it already has the same content of our "readthedoc" so... you know...

**Reply by fballiano** [reply]
> Very first ... "awesome openmage" should be linked and [updated](https://github.com/flagbit/Magento-ChangeAttributeSet).

@sreichel what do you mean? that the ChangeAttributeSet should be in the awesomeopenmage page? I can add it for sure, but it's not even linked in our readthedoc...

**Reply by sreichel** [reply]
I just wanted to have it link in README. :)

ChangeAttributeSet was just a suggestion.

**Comment by addison74** [comment]
Let's detach OpenMage from Magento 1 and promote it as an alternative platform to Magento 2, OpenCart, PrestaShop, Woo Commerce. Magento 1 is dead and buried and OpenMage is not Magento LTS. I hope that those who want to switch to M2 will think twice, having OpenMage as an alternative. Why change something that works, just for the sake of keeping up with the news? However, M2 has over 1300 reported issues and 810 PRs at this moment.

Let's convince the developers that OpenMage can offer them a way to continue making money, with the platform being updated. They can replace their Magento 1 sections with OpenMage. The Internet is full of solutions, books, tutorials, all related to Magento 1, it is a shame to waste them.

To solve all reported issues and PRs that are open. If we don't succeed, at least let's sort them. It is not difficult if we do this in a team if 6 - 10 people actively participate. I proposed myself to visit this repository for 1 hour a day.

Let's settle on two versions of OpenMage, one LTS to guarantee some bug fixes and MAIN to be a modern version with changes that cannot be made in the LTS version.

To update some Magento modules. For example, Mage_Captcha needs a better user interface similar to reCaptcha.

When all this is done we can make a TODO list in which to gather improvements in functionality. As @elidrissidev says above, the possibility to edit an order, it is true that they are extensions, but I am sure that some of you have a solution for this. I have a not very long list of changes that bring more functionality, for example, I recently ran into a situation where I had to disable a customer's account until a dispute was resolved. As well as having the possibility to hide temporary a custom option in Frontend, an additional checkbox to what exists now.

**Comment by fballiano** [comment]
We did a lot in 2022, solving bugs, introducing a lot of improvements, cleaning up extremely outdated code, moving to a more modern stack and I'm pretty sure we already have the best set of workflow for code quality of any other PHP project :-)
OM has some of the brightest minds of the ecommerce world and it is a privilege to to be part of it, all of this great work deserves so much more acknowledgment.

Moving to the difficult topics, here's my considerations.

I've always said that being solely a "LTS" means death, we need to be the best platform again, this is not possible if we get stuck on deleting a single file that nobody uses since forever because maybe that one person in the world (that is not updating OM anyway) could have a problem. We kill our whole project to save somebody who doesn't even care.

We have to acknowledge who we are: a community based project with less than 10 active people, thus:
- we cannot maintain 2 branches
- there is no reason why v19 should still support ie8
- if a user has to fix something while upgrading to a new version IT IS FINE (it doesn't mean let's destroy everything without caring, but not even wasting thousands of hours maintaining a v20 just because we don't want to remove ie8 from v19, because that's almost the only difference). Let's ask any M2 store how difficult is to upgrade a minor version, and let's compare to us.
- the team should be on a "turn" basis, it's not right that there are these many dormant roles in the team, maintainers that are not longer involved should step down, every role and privilege should have an expiration date, elections, meetings
- I'd like to see RFCs for the old decisions, for example who decided to have a website with that information? where's the RFC for that? who decided to provide 5years LTS? where's the RFC for that? I think those decision should be discussed every X time.
- who are these partners https://www.openmage.org/partners/index.html? why I never see anything from them here? For sure I'm missing something but that needs to be more transparent. I would need a couple more OM customers but on the website we are promoting in big letters companies that do not seem to help us that much. If they do, then I'm sorry but it needs to be more evident.
- we do not have the manpower the manage the website/readthedocs, we would be better off if the website would just redirect to the github, people would immediately see the amount to work that's done on a daily basis and the information shown on the website is mostly the same that's already in our README (apart from the debatable partners section)

The points above, IMO, would allow us to be more effective and to have a better working image for us (we're all developers at the end of the day).

From a software point of view I think we should have a full page cache functionality, we're lacking a modern varnish and that's already quite a problem. Our RWD theme should be modernized a little bit (it's mostly very good but could be improved a little bit). We'd need more "improved navigation" (everybody needs what I'm talking about) support, a little bit more advanced SEO features and we've to keep up with PHP 8.2. If we manage to be the first project to keep updated with the fast pace of PHP we will get people attention. 

You all deserve the best year ever!

**Reply by sreichel** [reply]
> we cannot maintain 2 branches

Agree.

> there is no reason why v19 should still support ie8

Agree.

> if a user has to fix something while upgrading to a new version IT IS FINE

Agree, but we should avoid this if possible. (Yeah,for sure,  some improvements may need some updates.)

> the team should be on a "turn" basis, it's not right that there are ...

Agree. 

From my POV ... i'm contributing since years, but have less priveleges then you (no problem, just to mention). I asked for, but no answer .... Also suggested @kiatng as maintainer for his code reviews ...

> who decided to provide 5years LTS

@Flyingmana only (?). Think this was fine, b/c it brought _some_ attention for this project!

> who are these partners ...

Really good question!

I can only match a few contributors to these partners.

It should be transparent why who is listed as a partner!

> we do not have the manpower the manage the website/readthedocs

Agree, but ...

> we would be better off if the website would just redirect to the github

It was really nice from @colinmollenhour (hope im correct) to build this website! And it is nice to have more then only a github repo. Problem is, that there is no documention to update it. (Adding/changing content, change html output, ...).

... and its outdated ...

- _[200 Merged Pull Requests](https://www.openmage.org/)_
- _[Meet our Mantainers](https://www.openmage.org/)_ Lee Saferite, David Robinson, Daniel Fahlke
- _PHP 7.4 Support_ ...
- working on full testsuite ...

**Reply by sreichel** [reply]
> I'm pretty sure we already have the best set of workflow for code quality of any other PHP project :-)

Not really. We have some tools to to follow some standards. We have no unit tests at all. PHPstan does not even pass level 0 w/o baseline. On a real modern project i'd target lvl8+ ... thats not possible w/o breaking most custom code that uses rewrites ...

Phpstan ... most changes related to missing or incorrect docblocks are fixed. Now, most baselined errors needs serious reviews (checking related code ...). Nothing i can do alone - or spending lot of time. So, there'll be only a few small phpstan updates next time. We also should fix baselined error in PHPCompatilibyCheck .... 

... I'll add a sponsorship button to my profile :rofl:

**Reply by Flyingmana** [reply]
> ... I'll add a sponsorship button to my profile 🤣

when you have, we can add it also to OpenMage repo for you

**Reply by fballiano** [reply]
> Not really. We have some tools to to follow some standards. We have no unit tests at all. PHPstan does not even pass level 0 w/o baseline. On a real modern project i'd target lvl8+ ... thats not possible w/o breaking most custom code that uses rewrites ...

ok, let's say "one of the best" :-) 

> Phpstan ... most changes related to missing or incorrect docblocks are fixed. Now, most baselined errors needs serious reviews (checking related code ...). Nothing i can do alone - or spending lot of time. So, there'll be only a few small phpstan updates next time. We also should fix baselined error in PHPCompatilibyCheck ....

we're still inheriting a very old codebase and we cannot do miracles (yet), I still think OM is at a very high level and has nothing to envy to anyone :-)

> ... I'll add a sponsorship button to my profile 🤣

that's where the partner companies (that do not participate in the project) should come in and that's why I've some critiques on the partners page on our website.

**Reply by fballiano** [reply]
> > if a user has to fix something while upgrading to a new version IT IS FINE
> 
> Agree, but we should avoid this if possible. (Yeah,for sure, some improvements may need some updates.)

absolutely, I never mean to be totally reckless :-)

> > the team should be on a "turn" basis, it's not right that there are ...
> 
> Agree.
> 
> From my POV ... i'm contributing since years, but have less priveleges then you (no problem, just to mention). I asked for, but no answer .... Also suggested @kiatng as maintainer for his code reviews ...

I'm super young in the admins team and, since there are many activities going on in the main repo, I still don't understand much about that :-D
you, elidrissidev and kiatng should definetely be in (in my opinion).
The "turn" obviously applies to me too, if one day I'll be no longer involved in OM, then I should be removed, or considered as removed.
I say this for some specific reason that has nothing to do with the people in the group (which are all great professionals).
When we take decisions we need a specific number of approvals, and that number has been decided (probably) based on the number of people who can vote. But if 7/10 admins are no longer responsive (it's just an example) that it's a big problem and can block everything. It also doesn't allow us to understand that we may need to introduce more new admins or that we can't handle certain activities in a certain way.
Again, nothing against the people. For example one of the admin is Vinai, which is one of my idols and inspirations, but is it really right that he is still in the admin group since (AFAIK) he's no longer involved in the project since long time?
Because of this discussion (I think this was the reason) one of the admin was removed from the group a couple of days ago, because of unresponsiveness.

I think we always have to question the status quo.

> > who decided to provide 5years LTS
> 
> @Flyingmana only (?). Think this was fine, b/c it brought _some_ attention for this project!

it was totally fine at that time but then we shouldn't answer "if you want to change something then RFC" when extremely important decisions were taken by a single person. otherwise those choices should be up for debate because they are still affecting everybody and situations constantly change

> It was really nice from @colinmollenhour (hope im correct) to build this website! And it is nice to have more then only a github repo. Problem is, that there is no documention to update it. (Adding/changing content, change html output, ...).

absolutely! but in 2022-2023 we're too small to maintain it and now the info is outdated and we have nobody keeping an eye on it. Technology changes too and maybe what was great 5 years ago could be an outdated way of thinking now, it doesn't invalidate what was made at that time.

**Reply by Flyingmana** [reply]
> I say this for some specific reason that has nothing to do with the people in the group (which are all great professionals).
> When we take decisions we need a specific number of approvals, and that number has been decided (probably) based on the number of people who can vote

It is not, its based on various standards of certification.

>  But if 7/10 admins are no longer responsive (it's just an example) that it's a big problem and can block everything.

All important Decisions are based on a minimum needed participation, and a possibly limited timeframe. (see rfc voting) Admins/Maintainers who do not participate during this timeframe, are not counted.


> it was totally fine at that time but then we shouldn't answer "if you want to change something then RFC" when extremely important decisions were taken by a single person. otherwise those choices should be up for debate because they are still affecting everybody and situations constantly change

yes, thats why I introduced the RFCs, so its not anymore "a single person decides", and also that its not hidden in some tickets or chats somewhere.

**Reply by fballiano** [reply]
> It is not, its based on various standards of certification.

If there's the need for 3 voters and in the group only 3 people ever participate, it means having a 100% adherence. You see what I mean and what's the problem here?

**Reply by Flyingmana** [reply]
> > It is not, its based on various standards of certification.
> 
> If there's the need for 3 voters and in the group only 3 people ever participate, it means having a 100% adherence. You see what I mean and what's the problem here?

its a problem which can not be solved by reducing the size of the group.

**Reply by fballiano** [reply]
but it will be made evident, when now it's not and we think it's all ok when it's not

**Reply by Flyingmana** [reply]
> * the team should be on a "turn" basis, it's not right that there are these many dormant roles in the team, maintainers that are not longer involved should step down, every role and privilege should have an expiration date, elections, meetings

I dont know how much it is going to help with it, but I started to list all the roles and responsibilities we have here: https://github.com/OpenMage/organizational/blob/main/Roles_and_Repsonsibilities.md

**Comment by kiatng** [comment]
To be competitive and attractive for _new_  projects, we need to remove legacy js libraries and have one of these: alpinejs, vuejs, **quasar**, reactjs, or some other modern js lib.

For existing projects, drop v19, continue with v20 as main.

**Reply by fballiano** [reply]
if you want v20 as main please vote here https://github.com/OpenMage/rfcs/pull/6

I'd love to see the legacy js gone, although I don't love the new js frameworks :-\

**Reply by elidrissidev** [reply]
So there's 2 PRS (yours and Dan's) that are proposing essentially the same thing, should we keep only one?

**Reply by Flyingmana** [reply]
> So there's 2 PRS (yours and Dan's) that are proposing essentially the same thing, should we keep only one?

not really, this one is about branch names and which one is the default branch, while mine is more about Release and support schedules. I would go so far and say they are in big parts complementary.

**Reply by addison74** [reply]
Quasar, AlpineJS, Tailwind are good options for creating an interface to fall in love with. Bootstrap 5.3 comes with new improvements. If we manage to replace the legacy with modern alternatives it would be a great achievement, but realistically speaking it is a lot of work. On the other hand, I don't even want to evaluate the impact on the extensions that use them. For example, the Color Picker from Magento 1 can be replaced with better good looking variants, but I have two extensions that use it and the developers don't want to hear about Magento 1 anymore. If our OpenMage platform will start to matter in production and the interest will increase with surely they will change their opinion. It's wonderful to see how much things have changed, I didn't think I would end up programming in Windows again, but DDEV radically changed my perspective.

I would have a proposal to fork Nexcess/Turpentine and connect it to OpenMage to start providing a solution for those who want to use Varnish. In the past I was close to this project. Thus we try to reduce the "great advantages" that Magento 2 has compared to Magento 1.

**Reply by fballiano** [reply]
@elidrissidev agree with @Flyingmana here, I'd keep them separated to have them as small as possible :-)

**Reply by fballiano** [reply]
@ADDISON74 agree, it's a very tough decision (and development). Regarding Turpentine, if would be great to have a working Varnish solution, I'm pretty sure @luigifab's version is the best nowadays https://github.com/luigifab/openmage-turpentine-varnish

**Reply by addison74** [reply]
@fballiano - @luigifab made a fork at another fork. Jeroen Vermeulen was a major contributor to Turpentine in a while. I have nothing against using this repository from now on for a Varnish solution.

**Comment by Flyingmana** [comment]
Adding this, because it was not mentioned yet:
Make it easier for new Developers and Users to learn how to work with OpenMage, by writing beginner guides and extend the general Documentation.

**Reply by luigifab** [reply]
Last week I found the [Magento Extension Developer Guide](https://info2.magento.com/rs/magentosoftware/images/magento-extension-developers-guide-v1.0.pdf).

**Reply by fballiano** [reply]
Wow lets save it somewhere before it disappears

**Reply by addison74** [reply]
Take this one too https://blog.magestore.com/free-magento-e-book-magento-made-easy-comprehensive-guide-to-magento-setup-and-development. 

Also, PacktPub sells a lot of nice books and videos related to Magento 1 https://subscription.packtpub.com/search?query=magento&page=1. The price per book is today $5.

**Reply by addison74** [reply]
There are a lot of Playlists on YouTube for users and developers.

**Reply by elidrissidev** [reply]
@luigifab good resource. I remember reading it when started learning how to create extensions.

**Reply by fballiano** [reply]
I've added these resources to https://github.com/fballiano/awesome-openmage, I just didn't add https://blog.magestore.com/free-magento-e-book-magento-made-easy-comprehensive-guide-to-magento-setup-and-development/ because when you click the "download" button then the ebook is no longer available.

**Reply by addison74** [reply]
I'm sorry I didn't check the link, I have that brochure in PDF format. I will add it as a link where it can be downloaded.

**Reply by sreichel** [reply]
https://alanstorm.com/category/magento/

**Reply by addison74** [reply]
This is a good resource for learning https://www.magestore.com/blog/magento-tutorial-for-beginners/

**Reply by fballiano** [reply]
> This is a good resource for learning https://www.magestore.com/blog/magento-tutorial-for-beginners/

@ADDISON74 great resource, added it to the awesome openmage

**Reply by justinbeaty** [reply]
Also from Alan Storm: https://alanstorm.com/series/checking-in-on-openmage-and-magento-in-2020/

**Comment by sreichel** [comment]
_Later_ ...

Moderize OM - modulize OM.

We removed `Mage_PageCache`, `Mage_Backup` and allowed to re-install them via composer. And there are already repos prepared for modules that are disabled by default - `Mage_Poll` & `Mage_Tag`). Old themes will move to git repo. Old MySql4 classes too ...

Why not add a repo for every Mage _Module_?

(i already tested this for some parts ...i works. We just need to have code in one repo for contributions (like M2) ...

**Reply by sreichel** [reply]
I know this will not come in near future, but it w'd be nice to have. Patches could be applied in the same way we do it for ZF1F.

**Comment by m-overlund** [comment]
Basic **online marketing** setup can be a hazzle in OM (pixels, serverside tracking etc.).
Things like (Google) Tag Manager(s), Google Analytics 4, Facebook Conversion API and the likes are mandatory for most webshops, and much easy to setup on the "competing" platforms.

OM already supports Google API / Universal Analytics for example.
- should follow through and be updated to support GA4 IMHO.

Whether you're an Ad Agency or Webshop owner, this is one of the issues i hear people (including myself) is dealing with.
WooCommerce and Shopify is often recommended in favor of OpenMage, in these areas.

**Reply by addison74** [reply]
Related to GA4, I brought up in the discussion that this year the OM code must be updated. I know that some can insert their own manual code, but we have this facility in the Backend interface. If it will not be updated it must be removed. The rest of the issues could be solved by extensions, but no developer has invested time in updating them. What I have analyzed recently, the main players have the last updates from 2018 - 2019, none for PHP 8.

**Reply by m-overlund** [reply]
I do agree on updating or removing Analytics support from OM.

For reference the only GA4 module I've found and is using is: https://www.scommerce-mage.com/magento-google-tag-manager.html
(Dunno if it supports PHP 8).

**Reply by sreichel** [reply]
I'd add GA4 support ... just need some sponsors ;)

**Comment by colinmollenhour** [comment]
Regarding the public website, the push for this really came from the EOL of M1 and the PCI compliance concerns that came with it. Some payment processors were positing that Magento 1 code was non-compliant and for shop owners to make the case to them that OpenMage is a viable migration path to maintain a secure installation it was very beneficial to have a more professional looking website. Unfortunately, the decision makers often don't understand what a GitHub project or pull request is and would not consider a GitHub README to be a "legitimate" online presence (unfortunately there is no objective standard for this).

I agree that it needs more maintenance so probably we should make it more maintenance-free by minimizing any information on it that is bound to become outdated. For example instead of "PHP 7.4 Support" it could say "Powered by the [actively supported](https://www.php.net/supported-versions) and [immensely popular](https://en.wikipedia.org/wiki/PHP#Popularity_and_usage_statistics) PHP project"

As far as documentation for the public site, it is basically a simple [Jekyll](https://jekyllrb.com/) site built by [GitHub Pages](https://pages.github.com/), nothing more. The content is stored in plain HTML/Markdown files so changing this verbiage is as simple as [finding the right file](https://github.com/OpenMage/OpenMage.github.io/search?q=merged) and submitting a PR. I just updated the README.md with this information.

All the same, thanks for bringing attention to this as it does need more attention. Would be great to get some community volunteers to scrutinize the entire site and submit some PRs.. What is the best way to seek such a volunteer currently?

**Reply by sreichel** [reply]
To work on this, there needs to be a better documentation.

- is it possible to aggregate some data from repo using github api? (e.g. to update commit counter)
- where does release information at http://127.0.0.1:4000/magento-lts/changelog/magento1-openmage-20_0.html come from - and how to improve output (better format for release information, or adding anchors?)

**Reply by fballiano** [reply]
In the meawhile I've updated some data https://github.com/OpenMage/OpenMage.github.io/pull/113

**Reply by sreichel** [reply]
PR for blog update .... https://github.com/OpenMage/OpenMage.github.io/pull/114

**Comment by addison74** [comment]
**Magento Community Tutorials Playlist**
https://www.youtube.com/watch?v=ktlGc6UWB9A&list=PL3B0BAAF482B16EAB

**Theming Magento Community Playlist**
https://www.youtube.com/watch?v=gsY51-QmxkU&list=PLLnpHn493BHHWhog4Xk36L1nyxUu16r9E

**Magento Training** 
https://www.youtube.com/playlist?list=PLcNSzkcgNPqSPyccfo8TIGZsUYtz81YpY
From 13 to 25 they are not free. See details here https://www.pierrefay.com/magento-training.html

**Magento Training Course**
https://www.youtube.com/playlist?list=PLa6k8rC4o6VXqTvFkBVm5iEwwm3gTqg74
https://www.youtube.com/@magentoeducationworld8388/playlists
https://www.youtube.com/watch?v=XAjxN5bbl0E&list=PLCi45QTgjDEvbHQm7u8mDzhrVgAphGpcP

**Magento Fundamentals**
https://www.youtube.com/playlist?list=PLa6k8rC4o6VWrk4sUVGjySuBoGUVa7J3t

**Magento Theme Creation**
https://www.youtube.com/@magentogodz2892/videos

**Template Monster Magento 1 & 2 Tutorials**
https://www.youtube.com/watch?v=kuvyFMYIiuM&list=PLhQIfRNfwAocfWc4zD0OPccvOuiO58-D6&index=2

**Mastering Magento - PacktPub**
https://www.youtube.com/@ShivaModi/videos

**Magento 1.9 - Let's CMS**
https://www.youtube.com/playlist?list=PLoTpPBZH6r134pilpxQJOor5L9cpkqrm3

**Magento Mechanics**
https://www.youtube.com/@codingmechanics2404/videos

**Magento Backend Administration**
https://www.youtube.com/watch?v=dR4lOUKnPgk
https://www.youtube.com/watch?v=386EQ4saV18
https://www.youtube.com/watch?v=Mg5TENnkYWY
https://www.youtube.com/watch?v=4tfUhXWMo_c
https://www.youtube.com/watch?v=BTEXEKrrvn0

**Reply by fballiano** [reply]
OMG and one is with Ben Marks!!!

added them to https://github.com/fballiano/awesome-openmage

**Reply by addison74** [reply]
... and now he is in the Shopware boat.

**Reply by fballiano** [reply]
well, he's been the face of the magento world for more than 10 years and somehow he'll always be a part of it :-)

**Comment by addison74** [comment]
**Quick Start to Magento Customization - Three Parts**
https://www.youtube.com/watch?v=I0bxZ6h4w78
https://www.youtube.com/watch?v=FGEa0TvdwLg
https://www.youtube.com/watch?v=g0VHo7B0YjY

These 3 videos will evoke nostalgia for some of you.
